### PR TITLE
[rpc-url-requirer] Made the relay-rpc-url functionality stateless. It…

### PR DESCRIPTION
**Pull request by awnns**
This PR fixes all problems we currently have with the rpc-url-requirer internal dictionary, apart from the hardest problem in programming.

### CODE REVIEW

**Comment by Maharacha** [comment]
Didn't know we have problems with it. But this is awesome anyway :+1:

**Comment by awnns** [comment]
> Didn't know we have problems with it. But this is awesome anyway 👍

I broke it a little because I had to perform deep magic to preserve the customer experience. It only affects certain situations, where the charm would throw exceptions when certain rare operations are performed and only if you have performed something resembling the deep magic I performed. I've been meaning to introduce this for a month or so to fix the problem, but Jonathan bumped into it yesterday, so I put in some effort.

**Comment by awnns** [comment]
> Nice! Did you test this on a node where we had this issue. E.g. `api-astar-rpc-16`

No, but I can do it later today. I tested it in a test environment on the lab cluster.

File: src/charm.py [hunk 1]
```diff
@@ -76,11 +76,19 @@ def __init__(self, *args):
...
-                                 relay_rpc_urls=dict())
+                                 )
+
+    def rpc_urls(self):
+        """
+        Return the rpc urls that are currently available in in all relay-rpc-url relations.
+        The magic comprehension below is the unfortunate actual best method to perform this
+        list flattening operation in Python.
+        """
+        return [subdata["ws_url"] for relation in self.model.relations["relay-rpc-url"] for subdata in relation.data.values() if "ws_url" in subdata]
```
Comment (jakobilobi) [inline src/charm.py:87]: Magic :D

**Review by jakobilobi — APPROVED** [review]
Approved on assumption you've tested it for more than one node

**Comment by Maharacha** [comment]
> > Nice! Did you test this on a node where we had this issue. E.g. `api-astar-rpc-16`
> 
> No, but I can do it later today. I tested it in a test environment on the lab cluster.

Status @awnns ?

**Comment by awnns** [comment]
> > > Nice! Did you test this on a node where we had this issue. E.g. `api-astar-rpc-16`
> > 
> > 
> > No, but I can do it later today. I tested it in a test environment on the lab cluster.
> 
> Status @awnns ?

I tested it on one of the backup nodes. It seemed worked, but there was a small hiccup during the switch to the new version iirc.

File: src/charm.py [hunk 2]
```diff
@@ -76,11 +76,19 @@ def __init__(self, *args):
                                  chain_spec_url=self.config.get('chain-spec-url'),
                                  local_relaychain_spec_url=self.config.get('local-relaychain-spec-url'),
                                  wasm_runtime_url=self.config.get('wasm-runtime-url'),
-                                 relay_rpc_urls=dict())
+                                 )
+
+    def rpc_urls(self):
+        """
+        Return the rpc urls that are currently available in in all relay-rpc-url relations.
```
Comment (jakobilobi) [inline src/charm.py:83]: Typo: in in

File: src/interface_rpc_url_provider.py [hunk 3]
```diff
@@ -11,7 +11,12 @@
 logger = logging.getLogger(__name__)
 
 class RpcUrlProvider(Object):
-    """RPC url provider interface."""
+    """
+    RPC url provider interface.
+
+    This interface is used by relay chain clients broadcast their rpc urls to parachain clients
+    that use them for relay-over-rpc.
```
Comment (jakobilobi) [inline src/interface_rpc_url_provider.py:18]: Nit: no consistency with capitalization of abbreviations (in more than this place)

```suggestion
    RPC URL provider interface.
    This interface is used by relay chain clients broadcast their RPC URLs to parachain clients
    that use them for relay-over-rpc.
```

**Review by jakobilobi — APPROVED** [review]
Hmm so we never actually need to access the relation changed event, only the relation data. Neat :+1:

**Pull request merged**